File: Backend/src/main/java/eu/echodream/medhead/Services/HopitalService.java
package eu.echodream.medhead.Services;

import eu.echodream.medhead.Models.Hopital;
import org.springframework.jdbc.core.JdbcTemplate;
import org.springframework.jdbc.core.RowMapper;
import org.springframework.stereotype.Service;

import java.util.List;

@Service
public class HopitalService {
    private final JdbcTemplate jdbcTemplate;
    private final SpecialisationService specialisationService;
    private final HopitalSpecialisationService hopitalSpecialisationService;

    public HopitalService(JdbcTemplate jdbcTemplate, SpecialisationService specialisationService,
                          HopitalSpecialisationService hopitalSpecialisationService) {
        this.jdbcTemplate = jdbcTemplate;
        this.specialisationService = specialisationService;
        this.hopitalSpecialisationService = hopitalSpecialisationService;
    }

    public List<Hopital> getAll() {
        String query = "SELECT Id, Nom, Adresse, Ville, CodePostal, LitsDisponibles FROM Hopital";
        List<Hopital> result = jdbcTemplate.query(query, rowMapper);
        result.forEach(hopital -> hopital.setSpecialisations(specialisationService.getByHopital(hopital.getId())));
        return result;
    }

    public Hopital get(int id) {
        String query = "SELECT Id, Nom, Adresse, Ville, CodePostal, LitsDisponibles FROM Hopital WHERE Id = ?";
        Hopital result = jdbcTemplate.queryForObject(query, new Object[]{id}, rowMapper);
        if (result != null) {
            result.setSpecialisations(specialisationService.getByHopital(id));
        }
        return result;
    }

    public int insert(Hopital hopital) {
        String query = "INSERT INTO Hopital (Nom, Adresse, Ville, CodePostal, LitsDisponibles) VALUES (?, ?, ?, ?, ?)";
        jdbcTemplate.update(query, hopital.getNom(), hopital.getAdresse(), hopital.getVille(),
                hopital.getCodePostal(), hopital.getLitsDisponibles());

        // get the last inserted id
        int id = jdbcTemplate.queryForObject("SELECT last_insert_rowid()", Integer.class);

        // On insère les associations avec les spécialisations
        hopital.getSpecialisations().forEach(specialisation -> hopitalSpecialisationService.insert(id, specialisation.getId()));

        return id;
    }

    public void update(Hopital hopital) {
        String query = "UPDATE Hopital SET Nom = ?, Adresse = ?, Ville = ?, CodePostal = ?, LitsDisponibles = ? WHERE Id = ?";
        jdbcTemplate.update(query, hopital.getNom(), hopital.getAdresse(), hopital.getVille(),
                hopital.getCodePostal(), hopital.getLitsDisponibles(), hopital.getId());

        // On supprime d'abord toutes les associations existantes
        hopitalSpecialisationService.deleteByHopital(hopital.getId());

        // On réinsère les nouvelles associations
        hopital.getSpecialisations().forEach(specialisation -> hopitalSpecialisationService.insert(hopital.getId(), specialisation.getId()));
    }

    public void delete(int id) {
        // On supprime d'abord toutes les associations existantes
        hopitalSpecialisationService.deleteByHopital(id);

        // On supprime l'hôpital
        String query = "DELETE FROM Hopital WHERE Id = ?";
        jdbcTemplate.update(query, id);
    }

    private static final RowMapper<Hopital> rowMapper = (rs, rowNum) ->
            new Hopital(rs.getInt("Id"), rs.getString("Nom"), rs.getString("Adresse"), rs.getString("Ville"),
                    rs.getString("CodePostal"), rs.getInt("LitsDisponibles"));
}

File: CSharp Version/EmergencyResponseSystem_DAL/Services/HopitalService.cs
using Dapper;
using EmergencyResponseSystem.Models;
using EmergencyResponseSystem_DAL.Database;
using EmergencyResponseSystem_Models;
using System;
using System.Collections.Generic;
using System.Linq;
using System.Text;
using System.Threading.Tasks;

namespace EmergencyResponseSystem_DAL.Services
{
    public class HopitalService
    {

        private readonly SpecialisationService _specialisationService;
        private readonly HopitalSpecialisationService _hopitalSpecialisationService;

        public HopitalService(SpecialisationService specialisationService, HopitalSpecialisationService hopitalSpecialisationService)
        {
            _specialisationService = specialisationService;
            _hopitalSpecialisationService = hopitalSpecialisationService;
        }

        public async Task<List<Hopital>> GetAllAsync()
        {
            using (var connection = DBManager.GetDBConnexion(true))
            {
                var query = "SELECT Id, Nom, Adresse, Ville, CodePostal, LitsDisponibles FROM Hopital";
                var result = await connection.QueryAsync<Hopital>(query);
                foreach (var hopital in result)
                {
                    hopital.Specialisations = await _specialisationService.GetByHopitalAsync(hopital.Id);
                }
                return result.ToList();
            }
        }

        public async Task<Hopital> GetAsync(int id)
        {
            using (var connection = DBManager.GetDBConnexion(true))
            {
                var query = "SELECT Id, Nom, Adresse, Ville, CodePostal, LitsDisponibles FROM Hopital WHERE Id = @Id";
                var result = await connection.QuerySingleOrDefaultAsync<Hopital>(query, new { Id = id });
                if (result != null)
                {
                    result.Specialisations = await _specialisationService.GetByHopitalAsync(id);
                }
                return result;
            }
        }

        public async Task<int> InsertAsync(Hopital hopital)
        {
            using (var connection = DBManager.GetDBConnexion(true))
            {
                var query = "INSERT INTO Hopital (Nom, Adresse, Ville, CodePostal, LitsDisponibles) VALUES (@Nom, @Adresse, @Ville, @CodePostal, @LitsDisponibles); SELECT last_insert_rowid();";
                var id = await connection.ExecuteScalarAsync<int>(query, hopital);

                // On insère les associations avec les spécialisations
                foreach (var specialisation in hopital.Specialisations)
                {
                    await _hopitalSpecialisationService.InsertAsync(hopital.Id, specialisation.Id);
                }

                return id;
            }
        }

        public async Task UpdateAsync(Hopital hopital)
        {
            using (var connection = DBManager.GetDBConnexion(true))
            {
                var query = "UPDATE Hopital SET Nom = @Nom, Adresse = @Adresse, Ville = @Ville, CodePostal = @CodePostal, LitsDisponibles = @LitsDisponibles WHERE Id = @Id";
                await connection.ExecuteAsync(query, hopital);

                // On supprime d'abord toutes les associations existantes
                await _hopitalSpecialisationService.DeleteByHopitalAsync(hopital.Id);

                // On réinsère les nouvelles associations
                foreach (var specialisation in hopital.Specialisations)
                {
                    await _hopitalSpecialisationService.InsertAsync(hopital.Id, specialisation.Id);
                }
            }
        }
        public async Task DeleteAsync(int id)
        {
            using (var connection = DBManager.GetDBConnexion(true))
            {
                // On supprime d'abord toutes les associations existantes
                await _hopitalSpecialisationService.DeleteByHopitalAsync(id);

                // On supprime l'hôpital
                var query = "DELETE FROM Hopital WHERE Id = @Id";
                await connection.ExecuteAsync(query, new { Id = id });
            }
        }

    }
}
